# **Maestría en Inteligencia Artificial Aplicada**

## Curso: **Procesamiento de Lenguaje Natural**

### Tecnológico de Monterrey

### Prof Luis Eduardo Falcón Morales

### **Actividad en Equipo - Semanas 4 y 5**

### **Vectores Embebidos de HuggingFace**

#### **Nombres y matrículas de los integrantes del equipo:**



- A01769076 | Alonso Pedrero Martínez
- A01797612 | Lilian Margarita Álvarez Iglesias
- A00000000 | Nombre


In [2]:
# Aquí deberán incluir todas las librerías que requieran durante esta actividad:

import pandas as pd
import numpy as np
import os
import re
from sklearn.model_selection import train_test_split
import string
import nltk
from collections import Counter
nltk.data.path.insert(0, os.path.abspath(os.path.join('..', '.venv', 'nltk_data')))


En esta actividad deberás utilizar los datos de tres archivos que se encuentran en el repositorio de la UCI llamados **amazon_cells_labelled.txt**, **imdb_labelled.txt** y   **yelp_labelled.txt**. Cada uno de estos archivos corresponden a comentarios de usuarios que adquirieron un celular a través de la plataforma de Amazon, de comentarios que dejaron usuarios sobre palículas y series en la plataforma de IMDb y sobre servicios de comida dejados en la plataforma de Yelp.

La información del problema y de los archivos están basados en el repositorio de la UCI cuya liga es la siguiente:

https://archive.ics.uci.edu/dataset/331/sentiment+labelled+sentences



# **Pregunta - 1:**



Descarga los 3 archivos de la plataforma de la UCI indicado previamente y genera un nuevo DataFrame de Pandas con ellos.

**Llama simplemente "df" a dicho DataFrame.**




In [3]:
PATH = "../files"

df_amazon = pd.read_csv(os.path.join(PATH, 'amazon_cells_labelled.txt'), sep='\t', names=['review','label'], header=None, encoding='utf-8')
dfi_imdb = pd.read_csv(os.path.join(PATH, 'imdb_labelled.txt'), delimiter='\t', names=['review','label'], header=None, encoding='utf-8')
df_yelp = pd.read_csv(os.path.join(PATH, 'yelp_labelled.txt'), sep='\t', names=['review','label'], header=None, encoding='utf-8')


In [4]:
df_amazon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   review  1000 non-null   object
 1   label   1000 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 15.8+ KB


In [5]:
dfi_imdb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 748 entries, 0 to 747
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   review  748 non-null    object
 1   label   748 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 11.8+ KB


In [6]:
df_yelp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   review  1000 non-null   object
 1   label   1000 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 15.8+ KB


In [7]:
newdfi=[]
dfi_list = dfi_imdb.values.tolist()

for row in dfi_list:

    review = str(row[0])
    label = str(row[1])

    text = review + '\t' + label

    matches = re.findall(r'(.*?)\t([01])(?:\n|$)', text, re.DOTALL)

    for rev, lab in matches:

        rev = rev.strip()

        if rev != '':
            newdfi.append([rev, int(lab)])

# convertir a DataFrame
clean_dfi_imdb = pd.DataFrame(newdfi, columns=['review', 'label'])

In [8]:
clean_dfi_imdb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   review  1000 non-null   object
 1   label   1000 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 15.8+ KB


In [9]:
df = pd.concat([df_amazon, clean_dfi_imdb, df_yelp], ignore_index=True) 
# *********** Aquí termina la sección de agregar código *************


In [10]:
# Verifiquemos la información del DataFrame:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   review  3000 non-null   object
 1   label   3000 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 47.0+ KB


In [11]:
# Y mostremos sus primeros registros:

df.head()

,review,label
0,So there is no way for me to plug it in here i...,0
1,"Good case, Excellent value.",1
2,Great for the jawbone.,1
3,Tied to charger for conversations lasting more...,0
4,The mic is great.,1


# **Pregunta - 2:**

Proceso de limpieza. Aplica el proceso de limpieza que consideres adecuado.











In [12]:
# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


# ************* Inicia la sección de agregar código:*****************************

X = df.review
y = df.label

assert X.shape == (3000,)
assert y.shape == (3000,)  

In [13]:
def clean_tok(doc):
  doc_cleaned = ''.join(
    char for char in doc.lower()
    if char not in string.punctuation and not char.isdigit())
  
  tokens = nltk.word_tokenize(doc_cleaned)
  
  tokens = [
    word for word in tokens
    if len(word) > 1]

  return tokens

In [14]:
Xclean = [clean_tok(x) for x in X]

# *********** Aquí termina la sección de agregar código *************

In [15]:
# Despleguemos los primeros comentarios después de tu proceso de limpieza:

for x in Xclean[0:5]:
  print(x)


['so', 'there', 'is', 'no', 'way', 'for', 'me', 'to', 'plug', 'it', 'in', 'here', 'in', 'the', 'us', 'unless', 'go', 'by', 'converter']
['good', 'case', 'excellent', 'value']
['great', 'for', 'the', 'jawbone']
['tied', 'to', 'charger', 'for', 'conversations', 'lasting', 'more', 'than', 'minutesmajor', 'problems']
['the', 'mic', 'is', 'great']


# **Pregunta - 3:**



Realicemos una partición aleatoria con los porcentajes que consideres más adecuados. Utiliza una semilla para su reproducibilidad.

In [16]:

# ************* Inicia la sección de agregar código:*****************************

Xtrain, x_val_and_test, Ytrain, y_val_and_test = train_test_split(Xclean, y, train_size=.70, shuffle=True, random_state=1)
Xval, Xtest, Yval, Ytest = train_test_split(x_val_and_test, y_val_and_test, test_size=.50, shuffle=True, random_state=1)

In [17]:
print('X,y Train:', len(Xtrain), len(Ytrain))
print('X,y Val:', len(Xval), len(Yval))
print('X,y Test', len(Xtest), len(Ytest))

# *********** Termina la sección de agregar código *************

X,y Train: 2100 2100
X,y Val: 450 450
X,y Test 450 450


# **Pregunta - 4:**




### **Construye tu vocabulario a continuación utilizando solamente el conjunto de Train:**


In [18]:
# a.	Usa el conjunto de entrenamiento para generar tu vocabulario
#     con un tamaño que consideres adecuado:


# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********

token_dict = Counter()

for k in range(len(Xtrain)):
  token_dict.update(Xtrain[k])


print('Longitud del diccionario:', len(token_dict))
print('\n(word,frequency):')
print(token_dict.most_common(10))


Longitud del diccionario: 4216

(word,frequency):
[('the', 1349), ('and', 753), ('is', 536), ('it', 485), ('to', 472), ('this', 449), ('of', 423), ('was', 413), ('in', 280), ('for', 242)]


In [19]:

min_freq = 5

midicc = {}

for token, freq in token_dict.items():

    if freq >= min_freq:
        midicc[token] = freq

# *********** Aquí termina la sección de agregar código *************

In [20]:
# b.	Indica el tamaño del vocabulario generado.

# ******* Inicia la sección de agregar código: ***********

print('Longitud del vocabulario generado:', len(midicc))

# *********** Aquí termina la sección de agregar código *************

Longitud del vocabulario generado: 646


In [21]:
# c.	Con el vocabulario generado, filtra los conjuntos de entrenamiento,
#     validación y prueba para que todos los comentarios usen solamente las
#     palabras de este vocabulario.

#     Llamar train_X, val_X y test_X a estos tres conjuntos.


# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


train_X = []
for ss in Xtrain:
  train_X.append([w for w in ss if w in midicc])

val_X = []
for ss in Xval:
  val_X.append([w for w in ss if w in midicc])

test_X = []
for ss in Xtest:
  test_X.append([w for w in ss if w in midicc])


# *********** Aquí termina la sección de agregar código *************


In [22]:
# Vemos el resultado de los primeros comentarios del conjunto de validación:

for ss in val_X[0:5]:
  print(ss)

['director', 'to', 'from', 'and', 'with', 'the', 'drama', 'and', 'comedy', 'you', 'can', 'never', 'tell', 'what', 'is', 'going', 'to', 'next']
['food', 'is', 'way', 'overpriced', 'and', 'are', 'small']
['and', 'good', 'prices']
['this', 'place', 'is', 'two', 'up']
['its', 'fun', 'fast', 'and', 'of', 'the', 'night', 'of']


# **Pregunta - 5:**

Incluye tus comentarios sobre cada modelo de HuggingFace indicado.

### ++++++++ Inicia la sección de agregar texto: +++++++++++

* **a) bge-base-en-v1.5**
- Es un modelo de embeddings desarrollado por Beijing Academy of Artificial Intelligence (BAAI).
- Soporta tareas de búsqueda semántica (retrieval), similitud textual y sistemas RAG.
- Fue entrenado principalmente para textos en inglés.
- Cuenta con 12 capas (layers).
- El tamaño del vector embebido es de 768 dimensiones.
- Tiene un buen equilibrio entre rendimiento, velocidad y consumo de memoria.


* **b) bge-large-en-v1.5**
- Es una versión más grande y precisa de la familia BGE.
- Está optimizado para tareas de recuperación de información y búsqueda semántica.
- Cuenta con 24 capas (layers).
- El tamaño del embedding es de 1024 dimensiones.
- Genera representaciones semánticas más precisas que la versión base.
- Requiere más memoria RAM/VRAM y mayor tiempo de procesamiento.

* **c)	e5-base-v2**
- Es un modelo de embeddings desarrollado por Microsoft.
- Está diseñado para tareas de recuperación semántica y comparación de textos.
- Cuenta con 12 capas (layers).
El tamaño del vector embebido es de 768 dimensiones.
- Funciona principalmente con textos en inglés.
- Los textos largos se truncan a un máximo de 512 tokens

### ++++++++ Termina la sección de agregar texto: +++++++++++

# **Pregunta - 6:**

In [23]:
# a) Cargar el modelo de embeddings de HuggingFace seleccionado:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


from sentence_transformers import SentenceTransformer

modelo_emb = SentenceTransformer('BAAI/bge-base-en-v1.5')

palabras_vocab = list(midicc.keys())
vectores_palabras = modelo_emb.encode(palabras_vocab, show_progress_bar=False)

diccionario_emb = {palabra: vector for palabra, vector in zip(palabras_vocab, vectores_palabras)}


# *********** Aquí termina la sección de agregar código *************

In [24]:
# b) Primeros 3 elementos clave:valor del diccionario generado.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


for palabra, vector in list(diccionario_emb.items())[:3]:
    print(f'{palabra}: {vector[:5]}...')


# *********** Aquí termina la sección de agregar código *************



his: [-0.01901011  0.03201383  0.02433453 -0.00354907  0.01592902]...
dont: [0.00957294 0.03321743 0.02284437 0.01627552 0.03065623]...
much: [-0.04203346 -0.02016109  0.01995522 -0.03272089 -0.00774416]...


In [25]:
# c) Tamaño del diccionario generado:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


print('Tamaño del diccionario:', len(diccionario_emb))


# *********** Aquí termina la sección de agregar código *************


Tamaño del diccionario: 646


# **Pregunta - 7:**




Generamos los vectores embebidos a partir de los conjuntos de entrenamiento, validación y prueba y con las características indicadas en el archivo PDF.

Los llamaremos trainEmb, valEmb y testEmb, respectivamente.


In [26]:
# a) Comentarios con vectores embebidos.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


def doc_a_vector(doc, dic):
    vectores = [dic[w] for w in doc if w in dic]
    if vectores:
        return np.mean(vectores, axis=0)
    return np.zeros(len(next(iter(dic.values()))))

trainEmb = np.array([doc_a_vector(doc, diccionario_emb) for doc in train_X])
valEmb   = np.array([doc_a_vector(doc, diccionario_emb) for doc in val_X])
testEmb  = np.array([doc_a_vector(doc, diccionario_emb) for doc in test_X])


# *********** Aquí termina la sección de agregar código *************

In [27]:
# b) Dimensiones de los conjuntos trainEmb, valEmb y testEmb.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


print('trainEmb:', trainEmb.shape)
print('valEmb:  ', valEmb.shape)
print('testEmb: ', testEmb.shape)


# *********** Aquí termina la sección de agregar código *************

trainEmb: (2100, 768)
valEmb:   (450, 768)
testEmb:  (450, 768)


# **Pregunta - 8:**

In [28]:
# Número de tokens generedos al obtener cada uno de los conjuntos trainEmb, valEmb y testEmb.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


tokens_train = sum(len([w for w in doc if w in diccionario_emb]) for doc in train_X)
tokens_val   = sum(len([w for w in doc if w in diccionario_emb]) for doc in val_X)
tokens_test  = sum(len([w for w in doc if w in diccionario_emb]) for doc in test_X)

print('Tokens en trainEmb:', tokens_train)
print('Tokens en valEmb:  ', tokens_val)
print('Tokens en testEmb: ', tokens_test)


# *********** Aquí termina la sección de agregar código *************

Tokens en trainEmb: 18185
Tokens en valEmb:   3874
Tokens en testEmb:  3546


# **Pregunta - 9:**



Entrenamiento y reporte de los modelos de Regresión Logística y Bosque Aleatorio (Random Forest).


In [29]:
# 9a) REGRESIÓN LOGÍSTICA:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

modeloLR = LogisticRegression(C=1.0, max_iter=1000, random_state=1)
modeloLR.fit(trainEmb, Ytrain)

acc_train_LR = accuracy_score(Ytrain, modeloLR.predict(trainEmb))
acc_val_LR   = accuracy_score(Yval,   modeloLR.predict(valEmb))

print(f'LR - Exactitud Train: {acc_train_LR:.4f}')
print(f'LR - Exactitud Val:   {acc_val_LR:.4f}')
print()
print(classification_report(Yval, modeloLR.predict(valEmb), target_names=['Negativo', 'Positivo']))


# *********** Aquí termina la sección de agregar código *************


LR - Exactitud Train: 0.7833
LR - Exactitud Val:   0.7644

              precision    recall  f1-score   support

    Negativo       0.75      0.76      0.76       218
    Positivo       0.77      0.77      0.77       232

    accuracy                           0.76       450
   macro avg       0.76      0.76      0.76       450
weighted avg       0.76      0.76      0.76       450



In [30]:
# 9b) BOSQUE ALEATORIO (Random Forest):

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


from sklearn.ensemble import RandomForestClassifier

modeloRF = RandomForestClassifier(n_estimators=200, max_depth=5, min_samples_leaf=5, random_state=1)
modeloRF.fit(trainEmb, Ytrain)

acc_train_RF = accuracy_score(Ytrain, modeloRF.predict(trainEmb))
acc_val_RF   = accuracy_score(Yval,   modeloRF.predict(valEmb))

print(f'RF - Exactitud Train: {acc_train_RF:.4f}')
print(f'RF - Exactitud Val:   {acc_val_RF:.4f}')
print()
print(classification_report(Yval, modeloRF.predict(valEmb), target_names=['Negativo', 'Positivo']))


# *********** Aquí termina la sección de agregar código *************

RF - Exactitud Train: 0.9086
RF - Exactitud Val:   0.7533

              precision    recall  f1-score   support

    Negativo       0.73      0.77      0.75       218
    Positivo       0.77      0.74      0.75       232

    accuracy                           0.75       450
   macro avg       0.75      0.75      0.75       450
weighted avg       0.75      0.75      0.75       450



# **Pregunta - 10**

**Proceso basado en modelos Preentrenados**

In [31]:
# 10a) Partición.:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


None


# *********** Aquí termina la sección de agregar código *************

In [32]:
# 10b) Vectores embebidos:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


None


# *********** Aquí termina la sección de agregar código *************

In [33]:
# 10c) REGRESIÓN LOGÍSTICA.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


None


# *********** Aquí termina la sección de agregar código *************

In [34]:
# 10d) BOSQUE ALEATORIO.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


None


# *********** Aquí termina la sección de agregar código *************

# **Pregunta - 11:**

In [35]:
# Reporte del mejor modelo y partición con el conjunto de Prueba.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


None


# *********** Aquí termina la sección de agregar código *************

# **Pregunta - 12:**



Incluye tus comentarios finales de la actividad.

### ++++++++ Inicia la sección de agregar texto: +++++++++++


### Preprocesamiento y vocabulario

- Se aplicó limpieza básica: conversión a minúsculas, eliminación de puntuación y dígitos, tokenización con NLTK y descarte de tokens de un solo carácter.
- Para construir el vocabulario se usó únicamente el conjunto de entrenamiento, con un umbral de **min_freq = 5**. Esto resultó en un vocabulario de **646 palabras**, eliminando términos muy raros que podrían introducir ruido sin aportar capacidad de generalización.
- La partición fue **70% entrenamiento / 15% validación / 15% prueba** con `random_state=1` para reproducibilidad, resultando en 2100, 450 y 450 muestras respectivamente.

### Modelo de embeddings

- Se eligió **bge-base-en-v1.5** (BAAI) por su buen equilibrio entre rendimiento semántico, velocidad de inferencia y consumo de memoria (768 dimensiones, 12 capas).
- Se generó un diccionario de embeddings `{palabra: vector}` a partir del vocabulario de 646 palabras. Cada documento se representó como el **promedio** de los vectores de sus tokens presentes en el diccionario.

### Hiperparámetros de los clasificadores

- **Regresión Logística:** `C=1.0` (regularización estándar), `max_iter=1000` para garantizar convergencia.
- **Bosque Aleatorio:** `n_estimators=200`, `max_depth=5`, `min_samples_leaf=5`. Los valores conservadores de profundidad y mínimo de muestras por hoja buscan reducir el sobreajuste, que es una tendencia común de RF con datos de texto de alta dimensionalidad.

### Resultados obtenidos (Q9)

| Modelo | Exactitud Train | Exactitud Val | Brecha |
|--------|----------------|---------------|--------|
| Regresión Logística | 78.33% | 76.44% | ~2% |
| Bosque Aleatorio    | 90.86% | 75.33% | ~15% |

- La **Regresión Logística** es el modelo más sólido: buena exactitud en validación y una brecha train-val de apenas ~2%, lo que indica buena generalización.
- El **Bosque Aleatorio** muestra sobreajuste notable (~15% de brecha) a pesar de los hiperparámetros conservadores (`max_depth=5`, `min_samples_leaf=5`), lo que sugiere que este enfoque de embeddings promediados no es el más adecuado para RF.
- Ambos modelos obtienen precisión y recall balanceados entre clases positiva y negativa (~0.75–0.77), lo que indica que no hay sesgo hacia ninguna clase.


### ++++++++ Termina la sección de agregar texto: +++++++++++

# **Fin de la Actividad de Vectores Embebidos - HuggingFace**